warden/x/act/keeper/actions.go

```diff
@@ -134,45 +136,74 @@ type actionCreatorKey struct{}
 // AddAction creates a new action.
 // The action is created with the provided creator as the first approver.
 // This function also tries to execute the action immediately if it's ready.
-func (k Keeper) AddAction(ctx context.Context, creator string, msg sdk.Msg, timeoutHeight uint64, expectedApproveExpression ast.Expression, expectedRejectExpression ast.Expression) (*types.Action, error) {
+func (k Keeper) AddAction(ctx context.Context, creator string, msg sdk.Msg, timeoutHeight uint64, expectedApproveExpression *ast.Expression, expectedRejectExpression *ast.Expression) (*types.Action, error) {
 	if err := k.validateActionMsgSigners(msg); err != nil {
 		return nil, err
 	}
 
 	ctx = ctxWithActionCreator(sdk.UnwrapSDKContext(ctx), creator)
-	ctx, rule, err := k.rulesRegistry.Get(ctx, msg)
+	ctx, approveRule, rejectRule, err := k.rulesRegistry.Get(ctx, msg)
 	if err != nil {
 		return nil, errors.Wrapf(types.ErrNoRuleRegistryHandler, "%v", err)
 	}
 
-	// todo: check that expressions from rulesRegistry (templateRegistry) match with expected
+	if approveRule.Expression.String() != expectedApproveExpression.String() {
+		return nil, types.ErrApproveExpressionNotMatched
+	}
+
+	if rejectRule.Expression.String() != expectedRejectExpression.String() {
+		return nil, types.ErrRejectExpressionNotMatched
+	}
 
 	wrappedMsg, err := codectypes.NewAnyWithValue(msg)
 	if err != nil {
 		return nil, err
 	}
 
 	ctxWithMsg := cosmoshield.NewContext(ctx, msg)
-	preprocessedExpr, mentions, err := k.preprocessRule(ctxWithMsg, rule)
+	preprocessedApproveExpr, approveMentions, err := k.preprocessRule(ctxWithMsg, approveRule)
+	if err != nil {
+		return nil, err
+	}
+
+	preprocessedRejectExpr, rejectMentions, err := k.preprocessRule(ctxWithMsg, rejectRule)
 	if err != nil {
 		return nil, err
 	}
 
+	mentions := approveMentions
+
+	approveMentionsSet := make(map[string]struct{})
+
+	for _, approveMention := range approveMentions {
+		approveMentionsSet[approveMention] = struct{}{}
+	}
+
+	for _, rejectMention := range rejectMentions {
+		_, exists := approveMentionsSet[rejectMention]
+		if !exists {
+			mentions = append(mentions, rejectMention)
+		}
+	}
+
 	// update the rule of this Action with the preprocessed expression
-	rule.Expression = preprocessedExpr
+	// todo: should be removed with removing Rule field in Action
+	approveRule.Expression = preprocessedApproveExpr
 
 	// create action object
 	timestamp := k.getBlockTime(ctx)
 	act := &types.Action{
-		Status:        types.ActionStatus_ACTION_STATUS_PENDING,
-		Approvers:     nil,
-		Rule:          rule,
-		Mentions:      mentions,
-		Msg:           wrappedMsg,
-		Creator:       creator,
-		TimeoutHeight: timeoutHeight,
-		CreatedAt:     timestamp,
-		UpdatedAt:     timestamp,
+		Status:            types.ActionStatus_ACTION_STATUS_PENDING,
+		Approvers:         nil,
+		Rule:              approveRule,
+		Mentions:          mentions,
```

Comment:
Do we really want to keep it as is? Why don't add another `RejectMentions`? Just curious

Reply:
Well, user can check his mention in both aprrove/reject rules itself, the fact that address is mentioned matters